src/deephyper/hpo/_search.py

```diff
@@ -47,6 +48,160 @@ def get_init_params_as_json(obj):
+                    for i in range(self.num_objective):
+                        result[f"objective_{i}"] = obj
+
+            # Add job.id
+            result["job_id"] = int(job.id.split(".")[1])
+
+            # Add job.status
+            result["job_status"] = job.status.name
+
+            # Profiling and other
+            # metadata keys starting with "_" are not saved (considered as internal/private)
+            metadata = {f"m:{k}": v for k, v in job.metadata.items() if k[0] != "_"}
+            result.update(metadata)
+
+            if hasattr(job, "pareto_efficient"):
+                result["pareto_efficient"] = job.pareto_efficient
+
+            jobs_as_dict.append(result)
+
+        return jobs_as_dict
+
+    def to_dataframe(self) -> pd.DataFrame:
+        df = pd.DataFrame(self._to_dict(self.jobs))
+        return df
+
+    def to_csv(self, path: str, partial: bool = False, flush: bool = False):
+        """Write results to CSV file.
+
+        Args:
+            path (str):
+
+                Path of the CSV file.
+
+            partial (bool, optional):
+
+                A boolean that indicates if the CSV write is only writing the last non-writen
+                chunk. Defaults to ``False``.
+
+            flush (bool, optional):
+
+                A boolean that indicates if the CSV write in the case where ``partial=True`` should
+                be flushed anyway. Otherwise if ``False`` it will not write to disk until there is
+                a successful job. Defaults to ``False``.
+        """
+        if not partial:
+            df = self.to_dataframe()
+            df.to_csv(path, index=False)
+            return
+
+        resultsList = self._to_dict(self.jobs[self._csv_cursor :])
+
+        if len(resultsList) > 0:
+            started_dumping = self._csv_cursor > 0
+            file_mode = "a" if started_dumping else "w"
+
+            if not (started_dumping):
+                for result in resultsList:
+                    # Waiting to start receiving non-failed jobs before dumping results
+                    is_single_obj_and_has_success = (
+                        "objective" in result and type(result["objective"]) is not str
+                    )
+                    is_multi_obj_and_has_success = (
+                        "objective_0" in result and type(result["objective_0"]) is not str
+                    )
+                    if is_single_obj_and_has_success or is_multi_obj_and_has_success or flush:
+                        self._csv_columns = result.keys()
+                        break
+
+            if self._csv_columns is not None:
+                with open(os.path.join(path), file_mode) as fp:
+                    writer = csv.DictWriter(fp, self._csv_columns, extrasaction="ignore")
+                    if not (started_dumping):
+                        writer.writeheader()
+                    writer.writerows(resultsList)
+                    self._csv_cursor += len(resultsList)
```

Comment:
Can this be avoided by converting the results list to a dataframe then use pandas to write the dataframe to csv file? This would eliminate importing the csv module.

Reply:
This logic is to avoid redundant writing. It just appends new lines. The pandas logic would re-write everything again an again.

Reply:
The `self.jobs = []` is a list of dictionaries, is that right? Each dictionary represents the results of a completed job, is that correct?

Reply:
The `self.jobs` is actually a list of `HPOJob` objects.